### ISSUE TITLE
Ordering of long lists of Kubernetes is sometimes odd

### ISSUE THREAD

**Issue by hausdorff** [pulumi/pulumi#1990]
Our Helm implementation employs a [`helmSort` function](https://github.com/pulumi/pulumi-kubernetes/blob/90d2513aabec6c58bfd7ef3bd855dcd54fc133bc/pack/nodejs/helm.ts#L109), which is intended to give a loose, approximate, and otherwise best-effort ordering over Kubernetes resources, so that things like `v1/Namespace`s are deployed before resources that go into them.

This ordering is occasionally completely ignored, but only for specific resources, and it's not yet clear to me why.

For due diligence, I've run in `--non-interactive` and also printed out the order in which `RegisterResource` is called, to confirm that we're sending them back to the engine in the right order, only to have it mess the order up later.

## The scenario

`ConfigMap` should almost always be registered and created before `Deployment`, per the `helmSort` implementation.

In the code example below, you will see that, in some cases, `ConfigMap` is reliably created last, despite being registered close to first place. This is particularly annoying when you have a very long list of resources, and your `Deployment` depends on that `ConfigMap`, but there are a lot of resources between them.

First, consider the following Kubernetes/AKS/Helm Chart program:

```typescript
import * as k8s from "@pulumi/kubernetes";
import { k8sCluster, k8sProvider } from "./cluster";

function fixChart(o: any) {
    // Fixes resource GroupVersion not found error.
    //
    // `apiVersion` is chosen based on `.Capabilities`, which is filled in by Tiller during a `helm
    // install`. Because we use `helm template`, the API server's `.Capabilities` are not available
    // to us, and we fall back to `v1alpha1`. But, this version of `prometheus-operator` only
    // installs `v1`, so we get an error. There is no place in the values.yaml to override this, so
    // here, we manually correct this problem.
    if (o && o.apiVersion == "monitoring.coreos.com/v1alpha1") {
        o.apiVersion = "monitoring.coreos.com/v1";
    }

    // Fixes validation error: spec.imagePullSecrets in body must be of type array: "null"
    if (
        o &&
        (o.kind == "Prometheus" || o.kind == "Alertmanager") &&
        o.spec &&
        o.spec.imagePullSecrets.length == 0
    ) {
        delete o.spec.imagePullSecrets;
    }
}

let operator = new k8s.helm.v2.Chart(
    "prometheus-operator",
    {
        repo: "coreos",
        chart: "prometheus-operator",
        version: "0.0.29"
    },
    { providers: { kubernetes: k8sProvider } }
);

const grafana = new k8s.helm.v2.Chart(
    "grafana",
    {
        repo: "coreos",
        chart: "grafana",
        version: "0.0.37",
        transformations: [fixChart]
    },
    { providers: { kubernetes: k8sProvider }, dependsOn: operator }
);

new k8s.helm.v2.Chart(
    "kube-prometheus",
    {
        repo: "coreos",
        chart: "kube-prometheus",
        version: "0.0.105",
        values: {
            // AKS does not expose etcd, the controller manager, or the scheduler through the
            // Kubernetes API. Delete the Prometheus exporter Services pointed at them.
            deployKubeScheduler: false,
            deployKubeControllerManager: false,
            deployKubeEtcd: false
        },
        transformations: [fixChart]
    },
    {
        providers: { kubernetes: k8sProvider },
        // Work-around for pulumi#1743. `dependsOn` does not wait for the children of a resource to
        // be complete, so we simply wait for the last resource in grafana to be booted up.
        dependsOn: grafana.getResource("v1/Service", "grafana-grafana")
    }
);

export let cluster = k8sCluster.name;
// export let kubeConfig = k8sCluster.kubeConfigRaw;
```

When I run this I get a very long list of resources, which I will abbreviate down to contain the `Deployment`s and `ConfigMap`s. You will notice that in the first chart, the `ConfigMap` appears near the top, as it should be. In the second two, it appears near the bottom, where it almost never should be.

This is pretty reliable. I run this program repeatedly, and the result is nearly always the same.

```
Previewing update of stack 'aks-helm'
Previewing changes:

     Type                                                           Name                                           Plan          Info
 *   pulumi:pulumi:Stack                                            azure-ts-aks-helm-aks-helm                     no change
 +   ├─ kubernetes:helm.sh:Chart                                    prometheus-operator                            create
 +   │  ├─ kubernetes:core:ConfigMap                                prometheus-operator                            create
[... 5 resources ...]
 +   │  ├─ kubernetes:apps:Deployment                               prometheus-operator                            create
[... 3 resources ...]
 +   ├─ kubernetes:helm.sh:Chart                                    grafana                                        create
[... 7 resources ...]
 +   │  ├─ kubernetes:extensions:Deployment                         grafana-grafana                                create
 +   │  └─ kubernetes:core:ConfigMap                                grafana-grafana                                create
 +   └─ kubernetes:helm.sh:Chart                                    kube-prometheus                                create
[... 45 resources ...]
 +      ├─ kubernetes:extensions:Deployment                         kube-prometheus-exporter-kube-state            create
 +      ├─ kubernetes:extensions:Deployment                         kube-prometheus-grafana                        create
[... 3 resources ...]
 +      ├─ kubernetes:core:ConfigMap                                kube-prometheus-grafana                        create
[... 3 resources ...]
```

I have no earthly idea what this would be, and I haven't had time to dig into it, but it definitely causes problems on longer Helm Charts, and it would be good to at least formulate a good understanding of why.

**Comment by joeduffy** [comment]
Also /cc @CyrusNajmabadi as he understands what controls the display ordering best. This really could just be a display issue in the progress code, possibly even by-design as part of the tree view.

**Comment by swgillespie** [comment]
I'll check this out today.

**Comment by CyrusNajmabadi** [comment]
The tree view keeps track of the order resource events were received in.  It then attempts to preserve that order among children for any particular resource.  That way people should never observe order of children changing in the list. They shoudl only observe items being added to the end.  

If you turn off the tree view, what order of events to you see things reported to the console with?

**Comment by lukehoban** [comment]
@hausdorff What exact guarantee is needed in this case?  The Pulumi engine today makes no guarantee that two consecutive `registerResource` calls will be run as CRUD operations on the provider in the same order they arrive.  And in fact in general, it cannot.  If a caller of `registerResource` needs to ensure and ordering dependency between two operations, they can specify `dependsOn` relationship between the resources.

For the case in question though, I'm unsure what the requirement is - is it:
1. The ConfigMap must start it's Create operation before the Deployment starts it's Create operation?
2. The ConfigMap must finish it's Create operation before the Deployment starts it's Create operation?
3. The ConfigMap and Deployment Create operations must be run concurrently to ensure that they both succeed?
4. Something else?

**Comment by hausdorff** [comment]
@lukehoban the `ConfigMap` just needs to not be dead last. If it appears before or approximately near the `Deployment`, we will in general have much more reproducible builds.

**Comment by lukehoban** [comment]
I'm still not sure I understand the requirement.  What is it exactly about "being dead last" that causes a problem?  Couldn't that same problem still occur if this was at some other position after some thing that depended on it?  It would help to get more precise here so we can design a correct-by-default way to handle this.

Importantly, Pulumi cannot currently attempt to make any guarantees about order unless ordering is enforced by `dependsOn`.  Simply registering a resources before another cannot be a guarantee (or even a suggestion really) that these operations against the backing provider will occur in this order.  We might be able to default more closely into that order, but I suspect it would lead to even more frustratingly unreliable deployments in the limit - and cause dependencies to be less correctly encoded in source programs.

**Comment by hausdorff** [comment]
> Importantly, Pulumi cannot currently attempt to make any guarantees about order unless ordering is enforced by dependsOn. Simply registering a resources before another cannot be a guarantee (or even a suggestion really) that these operations against the backing provider will occur in this order. 

So, backing up, the situation is:

- There is currently no way to change `dependsOn` in YAML and Helm chart resources.
- I'll speak to this issue more later in this response, but it does seem that `ConfigMap` is being pathologically scheduled dead last in reproducible conditions, and this does cause a handful of _very_ important, very large charts to fail out of the box. (Prometheus, grafana, kube-prometheus, and prometheus-operator are all affected.)

Let me be clear that I don't think we need to change any of our guarantees. I _am_ proposing that as a practical matter, we might need to iron out this pathological case.

> We might be able to default more closely into that order, but I suspect it would lead to even more frustratingly unreliable deployments in the limit

I too am sensitive to this, but if we were really committed to that approach then we'd randomize the ordering of resource creates that have no dependency information, as Go does when you enumerate keys in a `map`.

But this would be _disastrous_ for Helm and Kubernetes YAML, which is why I believe we should not.

> What is it exactly about "being dead last" that causes a problem?

If the `ConfigMap` is last in the list of creates, a `Deployment` that depends on it and is created ~halfway through the list (_i.e._, where they should be and where they usually are) could easily hang until timeout.

Usually this is because a bunch of `Service`s are hanging, waiting for `Deployment`s, which are in turn hanging waiting for the `ConfigMap` that appears dead last in the list.

So I see two options.

- Bump parallelism up to a degree such that it is extremely unlikely that any significant Helm chart will encounter this issue.
- Iron out the pathological behavior that causes `ConfigMap` to have been scheduled dead last 100% of the time. If the `ConfigMap` appeared even loosely in the ordering where it was expected to, it is vastly more likely that we would not encounter this situation, because of parallel creates.

On balance I prefer the second, because I doubt there's something special about `ConfigMap`, and the position I do not want to be in is that `Namespace` is pathologically scheduled last.

> Couldn't that same problem still occur if this was at some other position after some thing that depended on it? It would help to get more precise here so we can design a correct-by-default way to handle this.

Sure, because I'm not proposing we change the guarantees. But this would make it dramatically less likely. I'm not shooting for 100%, I'm shooting for "as good as `kubectl`."

**Comment by joeduffy** [comment]
I agree that if `ConfigMap` is pathologically and always scheduled last, we should understand that.

Are you sure, however, that it's not just a display issue? The way we draw the tree view is such that different resources will get reordered regularly. I'm not sure exactly what drives the ordering here.

Either way, it does seem like we're depending on delicate heuristics for something that we'd want to be robust. If you really want things to be scheduled in a particular order, we should be doing something to ensure this.

The naive approach would be to simply string all resources in a chart together, in the order in which they are encountered using `dependsOn`. My understanding is that this would not be right, because for some _prior_ resources to make forward progress in provisioning, _future_ resource creation requests need to be submitted. So, I really don't know what to suggest here, but it seems worrisome.

**Comment by hausdorff** [comment]
> Are you sure, however, that it's not just a display issue? The way we draw the tree view is such that different resources will get reordered regularly. I'm not sure exactly what drives the ordering here.

The behavior I observe is that it is rendered last, and then created last as well. This seems to indicate that it is an ordering bug.

**Comment by lukehoban** [comment]
> Bump parallelism up to a degree such that it is extremely unlikely that any significant Helm chart will encounter this issue.

This actually feels more like the right fix here.  It sounds like the "requirement" (to the extent one can be established here) is that everything in the chart be deployed in parallel.  If this requirement is not met, then subtle dependencies between things can cause hangs in Pulumi because we wait for completion before continuing.  

It was our plan to increase parallelism anyway - to unbounded parallelism by default assuming it causes no practical problems elsewhere.  Any reason why this is not a right and sufficient path to go down to address this?

**Comment by CyrusNajmabadi** [comment]
> Are you sure, however, that it's not just a display issue? The way we draw the tree view is such that different resources will get reordered regularly. I'm not sure exactly what drives the ordering here.

Just for informational purposes, here's how our ordering works:

1. We keep a monotonically increasing counter, and every time we *display* a resource for the first time, we assign it that counter's number, and then increase the counter.
2. Whenever we print the tree, we ensure that the children of any node are ordered using that counter.  This ensures that new children we display will show up will always order *after* existing children we've displayed.

This should ensure that there really is no reordering going on.  Now, you may see nodes show up anywhere in the tree.  However, in any given child-list, those nodes really should end up at the end of the list.  

--

Note: tehre may be bugs here.  But htat's at leas the intent of what the display code is trying to do.

**Comment by hausdorff** [comment]
> It sounds like the "requirement" (to the extent one can be established here) is that everything in the chart be deployed in parallel. [...] Any reason why this is not a right and sufficient path to go down to address this?

Yes. Some Kubernetes resources will fail if other resources are not deployed first. For example, `Namespace` must be deployed before anything that goes _in_ the `Namespace`. This is the reason why `kubectl` and `helm` both sort resource operation. The create operations are executed synchronously, but they look parallel because neither waits for the operation to complete.

Furthermore, I very seriously doubt this is an issue unique to `ConfigMap`, so it would be nice to know at least what is going on.

Increased parallelism will probably work better on average, though.

**Comment by joeduffy** [comment]
So, I was dying from curiosity, so I looked into this. I think this can be chalked up to normal async interleavings. I believe we've beaten a dead horse at this point, but just to reiterate, there are no guarantees about the order in which new resource registrations will be dispatched to the engine (`dependsOn` and dataflow dependencies aside), and there are no guarantees about the order in which the engine processes those registrations.

To narrow things down, I just looked at the `coreos/grafana` chart from @hausdorff's example.

Here is the resulting resource graph shown by the CLI:

![image](https://user-images.githubusercontent.com/3953235/46440832-7a041d80-c719-11e8-80d9-9f4622769491.png)

However, if we look, we will find out that the Helm provider registered resources in a different order:

```
PodSecurityPolicy
Secret
ConfigMap
ServiceAccount
ClusterRole
ClusterRoleBinding
Service
Deployment
ServiceMonitor
```

This obeys the `helmSort` routine's preferred ordering.

We see that indeed the synchronous invocation of the resource constructor (`NEWRES` below), and all of the pre-RPC logic inside of the runtime module (`PRERPC`), processed these registrations in order:

```
    NEWRES: t=kubernetes:extensions/v1beta1:PodSecurityPolicy, name=grafana-grafana
    PRERPC: t=kubernetes:extensions/v1beta1:PodSecurityPolicy, name=grafana-grafana
    NEWRES: t=kubernetes:core/v1:Secret, name=grafana-grafana
    PRERPC: t=kubernetes:core/v1:Secret, name=grafana-grafana
    NEWRES: t=kubernetes:core/v1:ConfigMap, name=grafana-grafana
    PRERPC: t=kubernetes:core/v1:ConfigMap, name=grafana-grafana
    NEWRES: t=kubernetes:core/v1:ServiceAccount, name=grafana-grafana
    PRERPC: t=kubernetes:core/v1:ServiceAccount, name=grafana-grafana
    NEWRES: t=kubernetes:rbac.authorization.k8s.io/v1:ClusterRole, name=psp-grafana-grafana
    PRERPC: t=kubernetes:rbac.authorization.k8s.io/v1:ClusterRole, name=psp-grafana-grafana
    NEWRES: t=kubernetes:rbac.authorization.k8s.io/v1:ClusterRoleBinding, name=psp-grafana-grafana
    PRERPC: t=kubernetes:rbac.authorization.k8s.io/v1:ClusterRoleBinding, name=psp-grafana-grafana
    NEWRES: t=kubernetes:core/v1:Service, name=grafana-grafana
    PRERPC: t=kubernetes:core/v1:Service, name=grafana-grafana
    NEWRES: t=kubernetes:extensions/v1beta1:Deployment, name=grafana-grafana
    PRERPC: t=kubernetes:extensions/v1beta1:Deployment, name=grafana-grafana
    NEWRES: t=kubernetes:monitoring.coreos.com/v1:ServiceMonitor, name=grafana
    PRERPC: t=kubernetes:monitoring.coreos.com/v1:ServiceMonitor, name=grafana
```

Unfortunately, from here, the ordering diverges. The RPCs are sent in this different order:

```
    RPC: t=pulumi:pulumi:Stack, name=k8s-prometheus-k8s-prometheus-dev
    RPC: t=kubernetes:helm.sh/v2:Chart, name=grafana
    RPC: t=kubernetes:core/v1:ServiceAccount, name=grafana-grafana
    RPC: t=kubernetes:core/v1:Secret, name=grafana-grafana
    RPC: t=kubernetes:rbac.authorization.k8s.io/v1:ClusterRoleBinding, name=psp-grafana-grafana
    RPC: t=kubernetes:rbac.authorization.k8s.io/v1:ClusterRole, name=psp-grafana-grafana
    RPC: t=kubernetes:core/v1:ConfigMap, name=grafana-grafana
    RPC: t=kubernetes:core/v1:Service, name=grafana-grafana
    RPC: t=kubernetes:monitoring.coreos.com/v1:ServiceMonitor, name=grafana
    RPC: t=kubernetes:extensions/v1beta1:PodSecurityPolicy, name=grafana-grafana
    RPC: t=kubernetes:extensions/v1beta1:Deployment, name=grafana-grafana
```

The `RegisterResource` RPCs are received in even a different order:

```
I1003 14:37:56.908845   78057 source_eval.go:614] ResourceMonitor.RegisterResource received: t=pulumi:pulumi:Stack, name=k8s-prometheus-k8s-prometheus-dev, custom=false, #props=0, parent=, protect=false, provider=, deps=[]
I1003 14:37:57.419917   78057 source_eval.go:614] ResourceMonitor.RegisterResource received: t=kubernetes:helm.sh/v2:Chart, name=grafana, custom=false, #props=4, parent=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::pulumi:pulumi:Stack::k8s-prometheus-k8s-prometheus-dev, protect=false, provider=, deps=[]
I1003 14:37:57.848389   78057 source_eval.go:614] ResourceMonitor.RegisterResource received: t=kubernetes:core/v1:Secret, name=grafana-grafana, custom=true, #props=5, parent=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart::grafana, protect=false, provider=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::pulumi:providers:kubernetes::default::04da6b54-80e4-46f7-96ec-b56ff0331ba9, deps=[]
I1003 14:37:57.848795   78057 source_eval.go:614] ResourceMonitor.RegisterResource received: t=kubernetes:rbac.authorization.k8s.io/v1:ClusterRole, name=psp-grafana-grafana, custom=true, #props=4, parent=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart::grafana, protect=false, provider=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::pulumi:providers:kubernetes::default::04da6b54-80e4-46f7-96ec-b56ff0331ba9, deps=[]
I1003 14:37:57.849195   78057 source_eval.go:614] ResourceMonitor.RegisterResource received: t=kubernetes:core/v1:ServiceAccount, name=grafana-grafana, custom=true, #props=4, parent=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart::grafana, protect=false, provider=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::pulumi:providers:kubernetes::default::04da6b54-80e4-46f7-96ec-b56ff0331ba9, deps=[]
I1003 14:37:57.850308   78057 source_eval.go:614] ResourceMonitor.RegisterResource received: t=kubernetes:monitoring.coreos.com/v1:ServiceMonitor, name=grafana, custom=true, #props=4, parent=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart::grafana, protect=false, provider=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::pulumi:providers:kubernetes::default::04da6b54-80e4-46f7-96ec-b56ff0331ba9, deps=[]
I1003 14:37:57.850701   78057 source_eval.go:614] ResourceMonitor.RegisterResource received: t=kubernetes:rbac.authorization.k8s.io/v1:ClusterRoleBinding, name=psp-grafana-grafana, custom=true, #props=5, parent=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart::grafana, protect=false, provider=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::pulumi:providers:kubernetes::default::04da6b54-80e4-46f7-96ec-b56ff0331ba9, deps=[]
I1003 14:37:57.850890   78057 source_eval.go:614] ResourceMonitor.RegisterResource received: t=kubernetes:extensions/v1beta1:PodSecurityPolicy, name=grafana-grafana, custom=true, #props=4, parent=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart::grafana, protect=false, provider=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::pulumi:providers:kubernetes::default::04da6b54-80e4-46f7-96ec-b56ff0331ba9, deps=[]
I1003 14:37:57.850988   78057 source_eval.go:614] ResourceMonitor.RegisterResource received: t=kubernetes:core/v1:Service, name=grafana-grafana, custom=true, #props=4, parent=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart::grafana, protect=false, provider=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::pulumi:providers:kubernetes::default::04da6b54-80e4-46f7-96ec-b56ff0331ba9, deps=[]
I1003 14:37:57.852925   78057 source_eval.go:614] ResourceMonitor.RegisterResource received: t=kubernetes:core/v1:ConfigMap, name=grafana-grafana, custom=true, #props=4, parent=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart::grafana, protect=false, provider=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::pulumi:providers:kubernetes::default::04da6b54-80e4-46f7-96ec-b56ff0331ba9, deps=[]
I1003 14:37:57.853473   78057 source_eval.go:614] ResourceMonitor.RegisterResource received: t=kubernetes:extensions/v1beta1:Deployment, name=grafana-grafana, custom=true, #props=4, parent=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart::grafana, protect=false, provider=urn:pulumi:k8s-prometheus-dev::k8s-prometheus::pulumi:providers:kubernetes::default::04da6b54-80e4-46f7-96ec-b56ff0331ba9, deps=[]
```

I assume this is because of some combination of our own Goroutine concurrency in the engine, mixed with the gRPC concurrency model. These RPCs of course complete in yet a different order:

```
I1003 14:38:00.151863   78057 eventsink.go:60] RegisterResource RPC finished: resource:k8s-prometheus-k8s-prometheus-dev[pulumi:pulumi:Stack]; err: null, resp: urn:pulumi:k8s-prometheus-dev::k8s-prometheus::pulumi:pulumi:Stack::k8s-prometheus-k8s-prometheus-dev,,,,
I1003 14:38:00.276799   78057 eventsink.go:60] RegisterResource RPC finished: resource:grafana[kubernetes:helm.sh/v2:Chart]; err: null, resp: urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart::grafana,,chart,,,grafana,repo,,,coreos,transformations,,,,,,,,,,,version,,,0.0.37,,
I1003 14:38:00.597374   78057 eventsink.go:60] RegisterResource RPC finished: resource:grafana-grafana[kubernetes:core/v1:Secret]; err: null, resp: urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart$kubernetes:core/v1:Secret::grafana-grafana,,apiVersion,,,v1,data,,,,,password,,,YWRtaW4=,user,,,YWRtaW4=,kind,,,Secret,metadata,,,,,labels,,,,,app,,,grafana,chart,,,grafana-0.0.37,heritage,,,Tiller,release,,,grafana,name,,,grafana-grafana,type,,,Opaque,,
I1003 14:38:00.626531   78057 eventsink.go:60] RegisterResource RPC finished: resource:psp-grafana-grafana[kubernetes:rbac.authorization.k8s.io/v1:ClusterRole]; err: null, resp: urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart$kubernetes:rbac.authorization.k8s.io/v1:ClusterRole::psp-grafana-grafana,,apiVersion,,,rbac.authorization.k8s.io/v1,kind,,,ClusterRole,metadata,,,,,labels,,,,,app,,,grafana,chart,,,grafana-0.0.37,heritage,,,Tiller,release,,,grafana,name,,,psp-grafana-grafana,rules,,,,,,,,,,apiGroups,,,,,,,,extensions,resourceNames,,,,,,,,grafana-grafana,resources,,,,,,,,podsecuritypolicies,verbs,,,,,,,,use,,
I1003 14:38:00.662391   78057 eventsink.go:60] RegisterResource RPC finished: resource:grafana-grafana[kubernetes:core/v1:ServiceAccount]; err: null, resp: urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart$kubernetes:core/v1:ServiceAccount::grafana-grafana,,apiVersion,,,v1,imagePullSecrets,,,,,,,kind,,,ServiceAccount,metadata,,,,,labels,,,,,app,,,grafana,chart,,,grafana-0.0.37,heritage,,,Tiller,release,,,grafana,name,,,grafana-grafana,,
I1003 14:38:00.693602   78057 eventsink.go:60] RegisterResource RPC finished: resource:grafana[kubernetes:monitoring.coreos.com/v1:ServiceMonitor]; err: null, resp: urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart$kubernetes:monitoring.coreos.com/v1:ServiceMonitor::grafana,,apiVersion,,,monitoring.coreos.com/v1,kind,,,ServiceMonitor,metadata,,,,,labels,,,,,app,,,grafana,chart,,,grafana-0.0.37,component,,,grafana,heritage,,,Tiller,prometheus,,,grafana,release,,,grafana,name,,,grafana,spec,,,,,endpoints,,,,,,,,,,interval,,,30s,port,,,http,jobLabel,,,grafana,namespaceSelector,,,,,matchNames,,,,,,,,default,selector,,,,,matchLabels,,,,,app,,,grafana,chart,,,grafana-0.0.37,,
I1003 14:38:00.721537   78057 eventsink.go:60] RegisterResource RPC finished: resource:psp-grafana-grafana[kubernetes:rbac.authorization.k8s.io/v1:ClusterRoleBinding]; err: null, resp: urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart$kubernetes:rbac.authorization.k8s.io/v1:ClusterRoleBinding::psp-grafana-grafana,,apiVersion,,,rbac.authorization.k8s.io/v1,kind,,,ClusterRoleBinding,metadata,,,,,labels,,,,,app,,,grafana,chart,,,grafana-0.0.37,heritage,,,Tiller,release,,,grafana,name,,,psp-grafana-grafana,roleRef,,,,,apiGroup,,,rbac.authorization.k8s.io,kind,,,ClusterRole,name,,,psp-grafana-grafana,subjects,,,,,,,,,,kind,,,ServiceAccount,name,,,grafana-grafana,namespace,,,default,,
I1003 14:38:00.750971   78057 eventsink.go:60] RegisterResource RPC finished: resource:grafana-grafana[kubernetes:extensions/v1beta1:PodSecurityPolicy]; err: null, resp: urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart$kubernetes:extensions/v1beta1:PodSecurityPolicy::grafana-grafana,,apiVersion,,,extensions/v1beta1,kind,,,PodSecurityPolicy,metadata,,,,,labels,,,,,app,,,grafana,chart,,,grafana-0.0.37,heritage,,,Tiller,release,,,grafana,name,,,grafana-grafana,spec,,,,,fsGroup,,,,,ranges,,,,,,,,,,max,,65535,min,,0,rule,,,MustRunAs,hostIPC,,,,false,hostNetwork,,,,true,hostPID,,,,true,hostPorts,,,,,,,,,,max,,65535,min,,0,privileged,,,,false,readOnlyRootFilesystem,,,,false,runAsUser,,,,,rule,,,RunAsAny,seLinux,,,,,rule,,,RunAsAny,supplementalGroups,,,,,ranges,,,,,,,,,,max,,65535,min,,0,rule,,,MustRunAs,volumes,,,,,,,,configMap,,,emptyDir,,,projected,,,secret,,,downwardAPI,,,persistentVolumeClaim,,,hostPath,,
I1003 14:38:00.787368   78057 eventsink.go:60] RegisterResource RPC finished: resource:grafana-grafana[kubernetes:core/v1:Service]; err: null, resp: urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart$kubernetes:core/v1:Service::grafana-grafana,,apiVersion,,,v1,kind,,,Service,metadata,,,,,labels,,,,,app,,,grafana,chart,,,grafana-0.0.37,heritage,,,Tiller,release,,,grafana,name,,,grafana-grafana,spec,,,,,ports,,,,,,,,,,name,,,http,port,,80,protocol,,,TCP,targetPort,,3000,selector,,,,,app,,,grafana,type,,,ClusterIP,,
I1003 14:38:00.829037   78057 eventsink.go:60] RegisterResource RPC finished: resource:grafana-grafana[kubernetes:core/v1:ConfigMap]; err: null, resp: urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart$kubernetes:core/v1:ConfigMap::grafana-grafana,,apiVersion,,,v1,data,,,,,deployment-dashboard.json,,,{
I1003 14:38:00.856625   78057 eventsink.go:60] RegisterResource RPC finished: resource:grafana-grafana[kubernetes:extensions/v1beta1:Deployment]; err: null, resp: urn:pulumi:k8s-prometheus-dev::k8s-prometheus::kubernetes:helm.sh/v2:Chart$kubernetes:extensions/v1beta1:Deployment::grafana-grafana,,apiVersion,,,extensions/v1beta1,kind,,,Deployment,metadata,,,,,labels,,,,,app,,,grafana,chart,,,grafana-0.0.37,heritage,,,Tiller,release,,,grafana,name,,,grafana-grafana,spec,,,,,replicas,,1,template,,,,,metadata,,,,,labels,,,,,app,,,grafana,release,,,grafana,spec,,,,,containers,,,,,,,,,,env,,,,,,,,,,name,,,GF_AUTH_BASIC_ENABLED,value,,,true,,,,,name,,,GF_AUTH_ANONYMOUS_ENABLED,value,,,true,,,,,name,,,GF_SECURITY_ADMIN_USER,valueFrom,,,,,secretKeyRef,,,,,key,,,user,name,,,grafana-grafana,,,,,name,,,GF_SECURITY_ADMIN_PASSWORD,valueFrom,,,,,secretKeyRef,,,,,key,,,password,name,,,grafana-grafana,image,,,grafana/grafana:5.0.0,name,,,grafana,ports,,,,,,,,,,containerPort,,3000,name,,,web,readinessProbe,,,,,failureThreshold,,10,httpGet,,,,,path,,,/api/health,port,,3000,periodSeconds,,1,successThreshold,,1,timeoutSeconds,,1,volumeMounts,,,,,,,,,,mountPath,,,/var/lib/grafana,name,,,grafana-storage,,,,,args,,,,,,,,--watch-dir=/var/grafana-dashboards,,,--grafana-url=http://127.0.0.1:3000,env,,,,,,,,,,name,,,GRAFANA_USER,valueFrom,,,,,secretKeyRef,,,,,key,,,user,name,,,grafana-grafana,,,,,name,,,GRAFANA_PASSWORD,valueFrom,,,,,secretKeyRef,,,,,key,,,password,name,,,grafana-grafana,image,,,quay.io/coreos/grafana-watcher:v0.0.8,name,,,grafana-watcher,volumeMounts,,,,,,,,,,mountPath,,,/var/grafana-dashboards,name,,,grafana-dashboards,serviceAccountName,,,grafana-grafana,volumes,,,,,,,,,,emptyDir,,,,,,name,,,grafana-storage,,,,,configMap,,,,,name,,,grafana-grafana,name,,,grafana-dashboards,,
```

We can assume the completions retire out of order due to engine parallelism and, indeed, if we pass `-p=1`, the resource ordering matches that of the outbound RPCs:

![image](https://user-images.githubusercontent.com/3953235/46441283-edf2f580-c71a-11e8-8a4f-0712fa60b206.png)

So, TL;DR, concurrency. I have no idea why the `ConfigMap` ends up taking more asynchronous turns than the other resources, but I'm not sure that question is particularly interesting. The reality is that everything is always going to happen out of order and our libraries will need to cope with it.

**Comment by joeduffy** [comment]
Reordering happens in all cases. The orders are just different

**Comment by joeduffy** [comment]
I saw this issue reported in the link Luke posts above. If this is truly required for correctness of the Kubernetes provider -- which it seems it is -- we will need to make some changes here.

I've been pondering this a bit and it seems that we could pursue a few different paths.

One is to guarantee the engine will dispatch resource registrations in the order in which they are received. The Kubernetes provider can then explicitly serialize all calls in the language host, and possibly do some sort of serialization in the Go provider itself to serialize outbound Kubernetes API calls. At least then we know the calls get made to Kubernetes in the right order.

Note that I worry about the possibility of deadlocks here, for inter-resource dependencies.

An alternarive is to, rather than make this guarantee in the engine, change the RPC interface to use begin/end calls. If we had this, I believe the Kubernetes provider would be able to orchestrate calls with the needed level of control. We considered that a long time ago for output properties, but at the time the engine didn't handle parallelism and so it was too big a change. It's probably easier now.

**Comment by hausdorff** [comment]
As I said in that issue, the key is that resource operations need to be started in approximately the correct order, but we don't need to wait for them to finish before continuing. I made a tentative recommendation, which I'm not super attached to, which is basically to allow collection component to specify a scheduling priority for immediate children. Importantly this is not a total order as depends on is-- it's an ordering on which operations should be started in which order.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by hausdorff** [comment]
That said, I think if the behavior described in this bug was approximately resolved I believe the bug would be essentially mitigated.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by joeduffy** [comment]
@pgavlin This is the issue I mentioned to you today.

**Comment by hausdorff** [comment]
After some consideration, we've agreed to try a simpler approach, which would make Pulumi much less vulnerable to ordering issues, and which seems likely to fix most manifestations of this bug. If we are wrong, we can re-open this bug and consider other options -- our goal will be to implement the simple solution quickly to make sure we have time to come back to this if we want to.

The solution will be occur in the Kubernetes provider, and in the Helm SDK, and involves:

* **Keep Helm's sort order.** Even though this does not guarantee resources will actually be created in that order, in general it increases the likelihood that we will have approximately the correct order.
* **Set default parallelism to be unbounded.** This is resolved in #2123.
* **Retry `create` REST requests if they fail or conflict.** We will _not_ retry the await logic. For example, if the API server rejects a resource because of the namespace doesn't exist, or the definition is not found, the REST request will be tried a few more times. This is similar to how `kubectl` handles the conflicts ([_e.g._](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/apply/apply.go#L94).
* **Add a little extra dependency information in the Helm SDK.** For example, because we know about `Deployment`'s schema, we can make sure that it depends on the `ConfigMap`s and `Secret`s that it references -- usnig a `.get` if they're not created by the Chart.

Because the current fix is likely to exist entirely in the provider and the Helm SDK, I will close this bug, since there is nothing left to do in pulumi/pulumi.

Let me know if you think I missed something.
